File: src/main/java/me/nasukhov/TukitaLearner/bot/command/AskQuestion.java
package me.nasukhov.TukitaLearner.bot.command;

import me.nasukhov.TukitaLearner.bot.io.Input;
import me.nasukhov.TukitaLearner.bot.io.Output;
import me.nasukhov.TukitaLearner.study.GroupQuestion;
import me.nasukhov.TukitaLearner.study.ProgressTracker;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Component("askQuestionCommand")
public class AskQuestion implements Handler {
    private static final String NO_MORE_QUESTIONS_LEFT = "У нас пока нет новых вопросов. Проверьте позже";
    private final ProgressTracker progressTracker;

    public AskQuestion(ProgressTracker progressTracker) {
        this.progressTracker = progressTracker;
    }

    @Override
    public boolean supports(Input input) {
        return input.isDirectCommand("ask");
    }

    @Override
    public void handle(Input input, Output output) {
        var channel = input.channel();
        Optional<GroupQuestion> result = progressTracker.createRandomForChannel(channel);

        if (result.isEmpty()) {
            // TODO share summary. reset progress
            output.write(NO_MORE_QUESTIONS_LEFT);

            return;
        }

        GroupQuestion newQuestion = result.get();

        Map<String, String> replies = new HashMap<>();
        int optionNum = 0;
        for (String replyVariant : newQuestion.listVariants()) {
            replies.put(replyVariant, String.format("qh answer %s %d", newQuestion.getId().toString(), ++optionNum));
        }

        output.promptChoice(newQuestion.getText(), replies);
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/command/AskQuestion.kt
package me.nasukhov.tukitalearner.bot.command

import me.nasukhov.tukitalearner.bot.io.Input
import me.nasukhov.tukitalearner.bot.io.Output
import me.nasukhov.tukitalearner.study.Group
import me.nasukhov.tukitalearner.study.GroupRepository
import me.nasukhov.tukitalearner.study.ProgressTracker
import org.springframework.stereotype.Component

@Component("askQuestionCommand")
class AskQuestion(
    private val progressTracker: ProgressTracker,
    private val groupRepository: GroupRepository,
) : Handler {
    companion object {
        private const val NO_MORE_QUESTIONS_LEFT = "У нас пока нет новых вопросов. Проверьте позже"
    }

    override fun supports(input: Input): Boolean = input.isDirectCommand("ask")

    override fun handle(
        input: Input,
        output: Output,
    ) {
        val group: Group = groupRepository.findById(input.channel.id).get()
        val result = progressTracker.createRandomForGroup(group)

        if (result.isEmpty) {
            // TODO share summary. reset progress
            output.write(NO_MORE_QUESTIONS_LEFT)

            return
        }

        val newQuestion = result.get()

        val replies = HashMap<String, String>()
        var optionNum = 0
        for (replyVariant in newQuestion.listVariants()) {
            replies.put(replyVariant, "qh answer ${newQuestion.id} ${++optionNum}")
        }

        output.promptChoice(newQuestion.text, replies)
    }
}
